src/Store.ts

```diff
@@ -0,0 +1,392 @@
+import * as aspect from 'dojo-core/aspect';
+import Evented from 'dojo-core/Evented';
+import { default as has, add as hasAdd } from 'dojo-core/has';
+import { EventObject, Handle, Hash } from 'dojo-core/interfaces';
+import * as lang from 'dojo-core/lang';
+import Promise from 'dojo-core/Promise';
+import Filter from './Filter';
+import * as dstore from './interfaces';
+import QueryMethod, { QueryMethodArgs } from './QueryMethod';
+
+// Detect __proto__, and avoid using it on Firefox, as they warn about
+// deoptimizations. The watch method is a clear indicator of the Firefox
+// JS engine.
+hasAdd('object-proto', Boolean((<any> {})['__proto__']) && !(<any> {}).watch);
+const hasProto = has('object-proto');
+
+export interface NewableStoreModel {
+	new (args?: {}): any;
+	createSubclass?: (arg: any[]) => any;
+	prototype: any;
+}
+
+export interface StoreArgs {
+	Model?: NewableStoreModel;
+}
+
+export default class Store<T> extends Evented implements dstore.Collection<T>, Hash<any> {
+	[ index: string ]: any;
+	/**
+	 * Creates an object, throws an error if the object already exists
+	 *
+	 * @param object The object to store.
+	 * @param directives Additional directives for creating objects.
+	 *
+	 * @return The object that was stored, with any changes that were made by
+	 * the storage system (like generated id)
+	 */
+	add: (object: any, directives?: dstore.PutDirectives) => any;
+
+	/**
+	 * Indicates if the events should automatically be fired for put, add, remove
+	 * method calls. Stores may wish to explicitly fire events, to control when
+	 * and which event is fired.
+	 */
+	autoEmitEvents: boolean = true;
+
+	fetch: (args?: dstore.FetchArgs) => dstore.FetchPromise<T>;
+
+	Filter: typeof Filter = Filter;
+
+	/**
+	 * Retrieves an object by its identity
+	 *
+	 * @param id The identity to use to lookup the object
+	 * @returns The object in the store that matches the given id.
+	 */
+	get: (id: dstore.StoreItem) => Promise<T> | void;
+
+	/**
+	 * Indicates the property to use as the identity property. The values of this
+	 * property should be unique.
+	 */
+	idProperty: string = 'id';
+
+	/**
+	 * This should be a entity (like a class/constructor) with a 'prototype' property that will be
+	 * used as the prototype for all objects returned from this store. One can set
+	 * this to the Model from dmodel/Model to return Model objects, or leave this
+	 * to null if you don't want any methods to decorate the returned
+	 * objects (this can improve performance by avoiding prototype setting),
+	 */
+	Model: NewableStoreModel;
+
+	/**
+	 * One can provide a parsing function that will permit the parsing of the data. By
+	 * default we assume the provide data is a simple JavaScript array that requires
+	 * no parsing (subclass stores may provide their own default parse function)
+	 */
+	parse: (...args: any[]) => any;
```

Comment:
Will this need to be converted to a method like the others?  If so, I believe the version in `Store` will just return what it is passed, based on the docs.